### PR TITLE
Allow duplicate identifiers across declarations.

### LINKED ISSUE #8675: Duplicate Identifier between node.d.ts and lib.d.ts
**TypeScript Version:** 

nightly with "node.d.ts" from typings

We recently add [this](https://github.com/Microsoft/TypeScript/blob/master/src/lib/es5.d.ts#L867)in to es5.d.ts
**Expected behavior:**
No error
**Actual behavior:** 
Duplicate Identifier with the same property defined in "node.d.ts"

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #8675

Duplicate identifiers *within the same block* are still disallowed. Additionally, we retain the restriction that the types of the declared symbols must be identical.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12893,6 +12893,91 @@ namespace ts {
...
+                        case SyntaxKind.NumericLiteral:
+                        case SyntaxKind.Identifier:
+                            memberName = (member.name as LiteralExpression | Identifier).text;
+                            break;
+                        default:
+                            continue;
+                    }
+
+                    if (hasProperty(names, memberName)) {
+                        error(member, Diagnostics.Duplicate_identifier_0, memberName);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: `member.name` for the error range.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12893,6 +12893,91 @@ namespace ts {
...
+                    }
+                }
+
+                const static = forEach(member.modifiers, m => m.kind === SyntaxKind.StaticKeyword);
+                const names = static ? staticNames : instanceNames;
+                switch (member.kind) {
+                    case SyntaxKind.Constructor:
+                        for (const param of (member as ConstructorDeclaration).parameters) {
+                            if (isParameterPropertyDeclaration(param)) {
+                                addName(names, param, (param.name as Identifier).text, property);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: `param.name` for the location.

File: src/compiler/checker.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: We report the duplicate declaration errors on all declarations, as ppl complained they could not find the other one.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -12893,6 +12893,91 @@ namespace ts {
...
+                        addName(names, member.name, memberName, property);
+                        break;
+                }
+            }
+
+            function addName(names: Map<number>, location: Node, name: string, meaning: number) {
+                if (hasProperty(names, name)) {
+                    const prev = names[name];
+                    if (prev & meaning) {
+                        error(location, Diagnostics.Duplicate_identifier_0, name);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: same comment as the other duplicate definition error later on.

File: tests/baselines/reference/binaryIntegerLiteralError.errors.txt [hunk 4]
```diff
@@ -16,13 +13,7 @@ tests/cases/conformance/es6/binaryAndOctalIntegerLiteral/binaryIntegerLiteralErr
     
     var obj1 = {
         0b11010: "hi",
-        ~~~~~~~
-!!! error TS2300: Duplicate identifier '0b11010'.
         26: "Hello",
-        ~~
-!!! error TS2300: Duplicate identifier '26'.
         "26": "world",
```
Comment (mhegazy) [inline tests/baselines/reference/binaryIntegerLiteralError.errors.txt]: this we should keep. it is an error under strict mode to have duplicate properties in an object literal.

**Comment by mhegazy** [comment]
do not see the handling for optionality.

File: src/compiler/checker.ts (hunk 1 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: I wonder if we need to keep doing that since now it's guaranteed to be in the same block

Comment (mhegazy) [inline src/compiler/checker.ts]: i see. yeah you are probably right.

File: tests/baselines/reference/binaryIntegerLiteralError.errors.txt (hunk 4 above)
Comment (RyanCavanaugh) [inline tests/baselines/reference/binaryIntegerLiteralError.errors.txt]: Still happens as long as there are no syntax errors in the file

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -12893,6 +12893,92 @@ namespace ts {
...
+                        case SyntaxKind.Identifier:
+                            memberName = (member.name as LiteralExpression | Identifier).text;
+                            break;
+                        default:
+                            continue;
+                    }
+
+                    if (hasProperty(names, memberName)) {
+                        error(member.symbol.valueDeclaration, Diagnostics.Duplicate_identifier_0, memberName);
+                        error(member, Diagnostics.Duplicate_identifier_0, memberName);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: member.name to get a better error span.

File: tests/baselines/reference/symbolProperty36.errors.txt [hunk 6]
```diff
@@ -1,13 +0,0 @@
-tests/cases/conformance/es6/Symbols/symbolProperty36.ts(2,5): error TS2300: Duplicate identifier '[Symbol.isConcatSpreadable]'.
-tests/cases/conformance/es6/Symbols/symbolProperty36.ts(3,5): error TS2300: Duplicate identifier '[Symbol.isConcatSpreadable]'.
-
-
-==== tests/cases/conformance/es6/Symbols/symbolProperty36.ts (2 errors) ====
-    var x = {
-        [Symbol.isConcatSpreadable]: 0,
```
Comment (mhegazy) [inline tests/baselines/reference/symbolProperty36.errors.txt]: should still be an error.

**Comment by mhegazy** [comment]
also can you add a test for computed properties
```ts
class C {
    ["a"]: 1,
    ["a"]: 2
}
```

**Comment by RyanCavanaugh** [comment]
Last call

**Pull request merged**